### PR TITLE
ci: Add Github Actions, code coverage

### PR DESCRIPTION
**Pull request by tmc**
Adds Github Actions for:

* Unit Tests
* Coverage reporting (including tagging of rediscache and sqlite test runs separately)
* ~~eslint Linting (only for edited files)~~

This has a number of benefits of the current CI configuration:
* Is faster (3m45s vs 6m20s in example run).
* Parallelizes test runs (rediscache, sqlite) as well as runs tests on other node versions (11.x, 12.x)
* ~~Shows inline lint warnings in files that have been edited (which should allow us to ratchet up eslint compliance).~~

Example pull request showing output:
https://github.com/tmc/Spoke-1/pull/3

~~Example PR showing lint violation:
https://github.com/tmc/Spoke-1/pull/4/files~~

Showing speed comparison:
Screen Shot 2020-04-26 at 12 06 15 PM

Code coverage output:
Screen Shot 2020-04-26 at 12 06 31 PM

Lint violation:
Screen Shot 2020-04-26 at 12 14 19 PM

### CODE REVIEW

**Comment by ibrand** [comment]
hey @tmc ! We have Travis set up running tests and I've never used github actions. Will this conflict and/or be redundant with a Travis setup?

**Comment by schuyler1d** [comment]
@ibrand can you review the lint rules here?

We need to remove the e2e tests which should probably be deleted -- they did not run consistently and may be 'dead' at this point.  @matteosb was thinking about moving toward Cypress.  Thoughts?

**Comment by tmc** [comment]
> hey @tmc ! We have Travis set up running tests and I've never used github actions. Will this conflict and/or be redundant with a Travis setup?

It will not conflict. It's partially redundant.

As far as I can tell travis isn't set up to report results on PRs which this will do. Reporting results is a better volunteer experience.

File: .github/workflows/lint.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      matrix:
+        node-version: [10.x, 11.x, 12.x]
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Node.js ${{ matrix.node-version }}
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+    - name: eslint
+      uses: reviewdog/action-eslint@v1
```
[the remark this answers is not in the document]
Comment (tmc) [reply .github/workflows/lint.yaml]: it uses eslint and will only report issues in files that have been edited.

**Comment by tmc** [comment]
> @ibrand can you review the lint rules here?
> 
> We need to remove the e2e tests which should probably be deleted -- they did not run consistently and may be 'dead' at this point. @matteosb was thinking about moving toward Cypress. Thoughts?

I can pull out the e2e tests.

Personally I find pupeteer lighter weight and easy to use.

**Comment by matteosb** [comment]
@tmc Travis is configured to run on every PR but reporting back to GitHub appears to have broken recently. I think @schuyler1d is looking into it.

RE: Pupeteer. I don't have experience with it but it looks pretty good, though maybe a little lower-level than Cypress? From a quick read of the documentation it looks like there are a number of pros and cons. If you are interested in getting involved with e2e testing I'd appreciate your feedback on the RFC:  https://github.com/MoveOnOrg/Spoke/issues/1494

**Comment by tmc** [comment]
Pushed a commit removing e2e test workflow.

**Review by schuyler1d — CHANGES_REQUESTED** [review]
I like this, and after confirming that Github Actions is free, since it's a public repository, I'd like to try this.

Can we change the lint action to only run the prettier command that we run with prettierrc and package.json's lint-staged settings?

We're getting into a mode where a lot of collaboration is happening all at once and we don't want PRs to stay clean without lint churn, so that reviews and merges are easier.

If there's a way to run the eslint options only for NEW files, I'm game for that.

**Comment by tmc** [comment]
@schuyler1d -- The lint action here should use the .eslintrc configuration in the root of the repository -- is there indication that it doesn't?

The reviewdog tool has configuration so it will only report check failures in files that have been edited since branching from origin/main.

**Comment by schuyler1d** [comment]
Right, but let's say I make an edit in lines from 10-12 in the file and then the eslint fixes lines 15-100.

1. It's hard to focus in review and see that 10-12 are the important parts
2. If another PR edits lines 50-60 then we get conflicts that are hard to merge.

If we get a clear moment to run eslint across the project then we can turn this on, but the last time we tried that eslint was causing miscompile errors in one section , so that's why we're where we are -- that's why I'd prefer disabled to deletion because we might be able to enable it in the future but right now let's stick with prettier which has stayed uptodate across changes.

**Comment by tmc** [comment]
> Right, but let's say I make an edit in lines from 10-12 in the file and then the eslint fixes lines 15-100.
> 
> 1. It's hard to focus in review and see that 10-12 are the important parts
> 2. If another PR edits lines 50-60 then we get conflicts that are hard to merge.
> 
> If we get a clear moment to run eslint across the project then we can turn this on, but the last time we tried that eslint was causing miscompile errors in one section , so that's why we're where we are -- that's why I'd prefer disabled to deletion because we might be able to enable it in the future but right now let's stick with prettier which has stayed uptodate across changes.

This change wouldn't *fix* lines X-U but would include warnings about them. I think this can serve as a nudge for authors to clean up files as they edit them.

**Comment by tmc** [comment]
@schuyler1d Pulled out linting. Opened https://github.com/MoveOnOrg/Spoke/issues/1527 for later treatment. PTAL

**Review by schuyler1d — APPROVED** [review]
:+1:

**Pull request merged**